File: src/go/kamada_kawai/kamada_kawai.go
package main

type Coord struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type Edge struct {
	Source int `json:"source"`
	Target int `json:"target"`
}

type KamadaKawai struct {
	N       int
	Coords  []Coord `json:"coords"`
	M       int
	Edge    []Edge `json:"edges"`
	D_ij    [][]int
	L_ij    [][]float64
	K_ij    [][]float64
	K       float64 `json:"K"`
	Epsilon float64 `json:"epsilon"`
	Display float64 `json:"display"`
}

func (kk *KamadaKawai) Init() {
	// Initialize N and M
	kk.N = len(kk.Coords)
	kk.M = len(kk.Edge)

	// Initialize D_ij
	kk.D_ij = make([][]int, kk.N)
	for i := 0; i < kk.N; i++ {
		kk.D_ij[i] = make([]int, kk.N)
	}
	for i := 0; i < kk.N; i++ {
		for j := 0; j < kk.N; j++ {
			if i != j {
				kk.D_ij[i][j] = kk.N + 1
			}
		}
	}
	for i := 0; i < kk.M; i++ {
		kk.D_ij[kk.Edge[i].Source-1][kk.Edge[i].Target-1] = 1
		kk.D_ij[kk.Edge[i].Target-1][kk.Edge[i].Source-1] = 1
	}
	// Floyd-Warshall for d_ij
	for k := 0; k < kk.N; k++ {
		for i := 0; i < kk.N; i++ {
			for j := 0; j < kk.N; j++ {
				if kk.D_ij[i][j] > kk.D_ij[i][k]+kk.D_ij[k][j] {
					kk.D_ij[i][j] = kk.D_ij[i][k] + kk.D_ij[k][j]
				}
			}
		}
	}

	var L_0 float64 = kk.Display

	// Get max from D_ij
	var max int = 0
	for i := 0; i < kk.N; i++ {
		for j := 0; j < kk.N; j++ {
			if kk.D_ij[i][j] > max {
				max = kk.D_ij[i][j]
			}
		}
	}

	var L float64 = L_0 / float64(max)

	// Initialize L_ij
	kk.L_ij = make([][]float64, kk.N)
	for i := 0; i < kk.N; i++ {
		kk.L_ij[i] = make([]float64, kk.N)
	}
	for i := 0; i < kk.N; i++ {
		for j := 0; j < kk.N; j++ {
			if i != j {
				kk.L_ij[i][j] = float64(kk.D_ij[i][j]) * L
			}
		}
	}

	// Initialize K_ij
	kk.K_ij = make([][]float64, kk.N)
	for i := 0; i < kk.N; i++ {
		kk.K_ij[i] = make([]float64, kk.N)
	}
	for i := 0; i < kk.N; i++ {
		for j := 0; j < kk.N; j++ {
			if i != j {
				kk.K_ij[i][j] = kk.K / (float64(kk.D_ij[i][j]) * float64(kk.D_ij[i][j]))
			}
		}
	}

}

File: src/julia/lib/kamada_kawai.jl
module Kamada_Kawai

export Coord, Edge, KamadaKawai

mutable struct Coord 
    x::Float64
    y::Float64
end # Coord

mutable struct Edge
    Source::Int
    Target::Int
end # Edge

mutable struct KamadaKawai
    n::Int
    coords::Vector{Coord}
    m::Int
    edges::Vector{Edge}

    d_ij::Matrix{Int}
    l_ij::Matrix{Float64}
    k_ij::Matrix{Float64}

    K::Float64
    epsilon::Float64
    display::Float64

    function get_d_ij(n::Int, coords::Vector{Coord}, m::Int, edges::Vector{Edge})
        d_ij = fill(n + 1, n, n)
        for i in 1:n
            d_ij[i, i] = 0
        end

        for i in 1:m
            source = edges[i].Source
            target = edges[i].Target
            d_ij[source, target] = 1
            d_ij[target, source] = 1
        end

        # Floyd Warshall algorithm
        for k in 1:n
            for i in 1:n
                for j in 1:n
                    if d_ij[i, j] > d_ij[i, k] + d_ij[k, j]
                        d_ij[i, j] = d_ij[i, k] + d_ij[k, j]
                    end
                end
            end
        end

        return d_ij
    end # get_d_ij

    function KamadaKawai(coords::Vector{Any}, edges::Vector{Any}, K::Float64, epsilon::Float64, display::Float64)
        n = length(coords)
        m = length(edges)

        # Convert the coords to Coord
        coords = [Coord(coords[i]["x"], coords[i]["y"]) for i in 1:n]

        # Convert the edges to Edge
        edges = [Edge(edges[i]["source"], edges[i]["target"]) for i in 1:m]

        # Compute the d_ij matrix
        d_ij = get_d_ij(n, coords, m, edges)

        # Get the max from d_ij
        max_d_ij = maximum(d_ij)

        # Compute the l_ij matrix
        L_0 = display
        L = L_0 / max_d_ij
        l_ij = d_ij .* L

        # Compute the k_ij matrix
        k_ij = K ./ (d_ij .^ 2)

        new(n, coords, m, edges, d_ij, l_ij, k_ij, K, epsilon, display)
    end # KamadaKawai

end # Kamada_Kawai
end # module